File: examples/full-local/game/src-tauri/src/matchmaking.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

#[derive(Serialize, Deserialize, Debug)]
pub struct MatchStatus {
    pub status: String,
    #[serde(rename = "roomId")]
    pub room_id: Option<String>,
    pub url: Option<String>,
}

pub struct MatchmakingClient {
    base_url: String,
    client: reqwest::Client,
}

impl MatchmakingClient {
    pub fn new(base_url: &str) -> Self {
        Self {
            base_url: base_url.to_string(),
            client: reqwest::Client::new(),
        }
    }

    pub async fn join_queue(
        &mut self,
        user_id: &str,
        display_name: &str,
        roster_config: Value,
        public_key: &str,
        signature: &str,
    ) -> Result<(), Box<dyn std::error::Error>> {
        let body = serde_json::json!({
            "userId": user_id,
            "displayName": display_name,
            "rosterConfig": roster_config,
            "publicKey": public_key,
            "signature": signature,
        });

        let response = self
            .client
            .post(format!("{}/join", self.base_url))
            .json(&body)
            .send()
            .await?;

        if !response.status().is_success() {
            let error_text = response.text().await?;
            return Err(format!("Failed to join queue: {}", error_text).into());
        }

        Ok(())
    }

    pub async fn check_status(
        &self,
        roster_hash: &str,
        public_key: &str,
        signature: &str,
    ) -> Result<MatchStatus, Box<dyn std::error::Error>> {
        let url = format!(
            "{}/status/{}?publicKey={}&signature={}",
            self.base_url, roster_hash, public_key, signature
        );

        let response = self.client.get(&url).send().await?;

        if !response.status().is_success() {
            let error_text = response.text().await?;
            return Err(format!("Failed to check status: {}", error_text).into());
        }

        let status: MatchStatus = response.json().await?;
        Ok(status)
    }

    pub async fn leave_queue(
        &self,
        roster_hash: &str,
        public_key: &str,
        signature: &str,
    ) -> Result<(), Box<dyn std::error::Error>> {
        let body = serde_json::json!({
            "rosterConfigHash": roster_hash,
            "publicKey": public_key,
            "signature": signature,
        });

        let response = self
            .client
            .post(format!("{}/leave", self.base_url))
            .json(&body)
            .send()
            .await?;

        if !response.status().is_success() {
            let error_text = response.text().await?;
            return Err(format!("Failed to leave queue: {}", error_text).into());
        }

        Ok(())
    }
}

File: examples/full-local/game/game-pwa/src/game/matchmaking.ts
import { SIGNATURE, createShaFromJSON } from "@roster-lock/utils";
import { RosterLockV1Config } from "@roster-lock/types";
import { CurrentUser } from "../context/GameSessionContext";

type PrivateKey = Parameters<typeof SIGNATURE.ASYMMETRIC.createSignature>[0];

export type MatchResult = { url: string; roomId: string };

const MAX_ATTEMPTS = 60;

// Ported from game-headless/src/steps/1-matchmaking.ts - plain fetch, so it
// was already browser-safe. Only real change: an onWaiting callback so the
// UI can show live status instead of headless's console logging.
export async function joinMatch(
  user: CurrentUser,
  rosterConfig: RosterLockV1Config,
  matchmakingUrl: string,
  onWaiting?: (attempt: number) => void,
): Promise<MatchResult> {
  const rosterHash = await createShaFromJSON(rosterConfig);
  await enterQueue(matchmakingUrl, user, rosterHash, rosterConfig);

  for (let attempt = 0; attempt < MAX_ATTEMPTS; attempt++) {
    const status = await checkStatus(matchmakingUrl, user, rosterHash);
    if (status.status === "matched") return { url: status.url, roomId: status.roomId };
    onWaiting?.(attempt);
    await new Promise((resolve) => setTimeout(resolve, 1000));
  }
  throw new Error("Timed out waiting for match");
}

async function enterQueue(
  matchmakingUrl: string,
  user: CurrentUser,
  rosterHash: string,
  rosterConfig: RosterLockV1Config,
) {
  const timestamp = Date.now();
  const signature = await SIGNATURE.ASYMMETRIC.createSignature(user.keys.privateKey as PrivateKey, {
    service: "join-queue",
    machineId: user.machineId,
    displayName: user.displayName,
    rosterHash,
    timestamp,
    publicKey: user.keys.publicKey,
  });

  const response = await fetch(`${matchmakingUrl}/join`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({
      machineId: user.machineId,
      displayName: user.displayName,
      rosterConfig,
      publicKey: user.keys.publicKey,
      timestamp,
      signature,
    }),
  });

  if (!response.ok) {
    throw new Error(`Failed to join matchmaking queue (${response.status})`);
  }
  return (await response.json()) as { status: string; position: number; queuedAt: number };
}

type Status = { status: "waiting" } | { status: "matched"; roomId: string; url: string };

async function checkStatus(matchmakingUrl: string, user: CurrentUser, rosterHash: string): Promise<Status> {
  const timestamp = Date.now();
  const signature = await SIGNATURE.ASYMMETRIC.createSignature(user.keys.privateKey as PrivateKey, {
    service: "queue-status",
    rosterConfigHash: rosterHash,
    timestamp,
    publicKey: user.keys.publicKey,
  });

  const url = new URL(`/status/${rosterHash}`, matchmakingUrl);
  url.searchParams.append("publicKey", user.keys.publicKey);
  url.searchParams.append("signature", signature);
  url.searchParams.append("timestamp", timestamp.toString());

  const response = await fetch(url);
  if (!response.ok) {
    throw new Error(`Failed to check match status (${response.status})`);
  }
  return (await response.json()) as Status;
}
